plugin.video.embycon/resources/lib/datamanager.py

```diff
@@ -0,0 +1,159 @@
+            os.makedirs(os.path.join(__addondir__, "cache"))
+        cacheDataPath = os.path.join(__addondir__, "cache", urlHash)
+
+        log.debug("Cache_Data_Manager:" + cacheDataPath)
+
+        # are we forcing a reload
+        WINDOW = HomeWindow()
+        force_data_reload = WINDOW.getProperty("force_data_reload") == "true"
+        WINDOW.clearProperty("force_data_reload")
+
+        if os.path.exists(cacheDataPath) and not force_data_reload:
+            # load data from cache if it is available and trigger a background
+            # verification process to test cache validity   
+            log.debug("Loading Cached File")
+            with open(cacheDataPath, 'r') as f:
+                result = self.loadJasonData(f.read())
+
+            # start a worker thread to process the cache validity
+            self.cacheDataResult = result
+            self.dataUrl = url
+            self.cacheDataPath = cacheDataPath
+            actionThread = CacheManagerThread()
+            actionThread.setCacheData(self)
+            actionThread.start()
+
+            log.debug("Returning Cached Result")
+            return result
+        else:
+            # no cache data so load the url and save it
+            jsonData = DownloadUtils().downloadUrl(url, suppress=False, popup=1)
+            log.debug("Loading URL and saving to cache")
+            with open(cacheDataPath, 'w') as f:
+                f.write(jsonData)
+            result = self.loadJasonData(jsonData)
+            self.cacheManagerFinished = True
+            log.debug("Returning Loaded Result")
+            return result
+
+
+class CacheManagerThread(threading.Thread):
+    dataManager = None
+
+    def __init__(self, *args):
+        threading.Thread.__init__(self, *args)
+
+    def setCacheData(self, data):
+        self.dataManager = data
+
+    def run(self):
+
+        log.debug("CacheManagerThread Started")
+
+        cacheValidatorString = self.dataManager.getCacheValidatorFromData(self.dataManager.cacheDataResult)
+        log.debug("Cache Validator String (" + cacheValidatorString + ")")
+
+        jsonData = DownloadUtils().downloadUrl(self.dataManager.dataUrl, suppress=False, popup=1)
+        loadedResult = self.dataManager.loadJasonData(jsonData)
+        loadedValidatorString = self.dataManager.getCacheValidatorFromData(loadedResult)
+        log.debug("Loaded Validator String (" + loadedValidatorString + ")")
+
+        # if they dont match then save the data and trigger a content reload
+        if (cacheValidatorString != loadedValidatorString):
+            log.debug("CacheManagerThread Saving new cache data and reloading container")
+            with open(self.dataManager.cacheDataPath, 'w') as f:
+                f.write(jsonData)
+
+            # we need to refresh but will wait until the main function has finished
+            loops = 0
+            while (self.dataManager.canRefreshNow == False and loops < 200):
```

Comment:
I would add yet another condition here in the case the user decides to abort/quit kodi while you are stuck in the loop:

```python
while (self.dataManager.canRefreshNow == False and loops < 200 and not xbmc.Monitor().abortRequested()):
```

This is particularly important since 200 cycles of 100 milliseconds will lead to a 20 seconds wait time. By default kodi tries to kill running sub-processes if they do not abort in a 5 seconds interval.